src/utils/format.ts

```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+    thousandsSymbol,
+    decimalSymbol,
+  }: CompactionOptions): string {
+  // e.g TRILLION_123_123.div(ONE_TRILLION) = 123123.123123123
+  const numAsFraction = baseBN.mul(TEN.pow(new BN(precision))).div(representationBN)
+  const { integerPart, decimalPart } = _decomposeBn(numAsFraction, DEFAULT_LARGE_NUMBER_PRECISION, DEFAULT_LARGE_NUMBER_PRECISION)
+  // 123123.123123123 = 123,123.123123123
+  const formattedInteger = _formatNumber(integerPart.toString(), thousandsSymbol)
+  // no relevant decimal section
+  if (decimalPart.isZero()) return formattedInteger
+
+  const formattedDecimal = decimalPart
+    .toString()
+    .padStart(precision, '0') // Pad the decimal part with leading zeros
+    .replace(/0+$/, '') // Remove the right zeros
+  return formattedInteger + decimalSymbol + formattedDecimal
+}
 
 function _formatNumber(num: string, thousandsSymbol: string = THOUSANDS_SYMBOL): string {
   return num.replace(/(\d)(?=(?:\d{3})+(?!\d))/g, '$1' + thousandsSymbol)
 }
 
-function _decomposeBn(amount: BN, amountPrecision: number, decimals: number): { integerPart: BN; decimalPart: BN } {
-  // Discard the decimals we don't need
-  //  i.e. for WETH (precision=18, decimals=4) --> amount / 1e14
-  //        16.5*1e18 ---> 165000
+interface DecomposedNumberParts {
+  integerPart: BN
+  decimalPart: BN
+  decimalsPadded: string
+}
+
+function _formatSmart(
+  { integerPart, decimalPart, decimalsPadded }: DecomposedNumberParts,
+  thousandsSymbol: string = THOUSANDS_SYMBOL,
+  decimalSymbol: string = DECIMALS_SYMBOL,
+  smallLimitAsWei: BN,
+  precision: number = DEFAULT_LARGE_NUMBER_PRECISION,
+): string {
+  // Is fraction
+  if (integerPart.isZero()) {
+    // if amount < 1 and fraction < smallLimit (both compared as Wei)
+    // return `< ${smallLimit}`
+    // else return decimals as is
+    // first we need to convert decimals to WEI in order to compare small values
+    const ourDecimalsAsBNWei = new BN(toWei('0.' + decimalsPadded))
+    return ourDecimalsAsBNWei.lt(smallLimitAsWei) ? `< ${_formatDecimalsFromWeiForDisplay(smallLimitAsWei, decimalSymbol)}` : _formatDecimalsFromWeiForDisplay(ourDecimalsAsBNWei)
+  }
+
+  const compactionOptions = { precision, thousandsSymbol, decimalSymbol }
+
+  // Number compacting logic
+  // Anything > 1,000,000,000,000 denoted as <XXX,XXX.xxx>T
+  if (integerPart.gte(ONE_TRILLION_BN)) {
+    return _compactLargeBNNumberToString(integerPart, ONE_TRILLION_BN, compactionOptions) + 'T'
+  }
+  // Anything not above 1T but > 1,000,000,000 denoted as <XXX.xxx>B
+  if (integerPart.gte(ONE_BILLION_BN)) {
+    return _compactLargeBNNumberToString(integerPart, ONE_BILLION_BN, compactionOptions) + 'B'
+  }
+
+  // At this point can just return thousands separated formatted integer
+  // if decimals are zero
+  if (decimalPart.isZero()) return _formatNumber(integerPart.toString())
+
+  let finalPrecision: number
+
+  // normal format + separator
+  if (integerPart.lt(ONE_HUNDRED_K)) {
+    // 100,000.3424
```

Comment:
The code doesn't match the comment.
100'000 not less than 100'000.
is this a bad example or should be less than equal?

Reply:
oops ur right, fixing